File: android/lib/src/main/java/com/swmansion/gesturehandler/RotationGestureDetector.java
package com.swmansion.gesturehandler;

import android.view.MotionEvent;

public class RotationGestureDetector {

  public interface OnRotationGestureListener {

    boolean onRotation(RotationGestureDetector detector);

    boolean onRotationBegin(RotationGestureDetector detector);

    void onRotationEnd(RotationGestureDetector detector);
  }

  private long mCurrTime;
  private long mPrevTime;
  private double mPrevAngle;
  private double mAngleDiff;

  private boolean mInProgress;

  private int mPointerIndices[] = new int[2];

  private OnRotationGestureListener mListener;

  public RotationGestureDetector(OnRotationGestureListener listener) {
    mListener = listener;
  }

  private float getRawX(MotionEvent event, int index) {
    float offset = event.getX() - event.getRawX();
    return event.getX(mPointerIndices[index]) + offset;
  }

  private float getRawY(MotionEvent event, int index) {
    float offset = event.getY() - event.getRawY();
    return event.getY(mPointerIndices[index]) + offset;
  }

  private void updateCurrent(MotionEvent event) {
    mPrevTime = mCurrTime;
    mCurrTime = event.getEventTime();

    float vectorX = getRawX(event, 1) - getRawX(event, 0);
    float vectorY = getRawY(event, 1) - getRawY(event, 0);

    // Angle diff should be positive when rotating in clockwise direction
    double angle = -Math.atan2(vectorY, vectorX);

    if (Double.isNaN(mPrevAngle)) {
      mAngleDiff = 0.;
    } else {
      mAngleDiff = mPrevAngle - angle;
    }
    mPrevAngle = angle;

    if (mAngleDiff > Math.PI / 2) {
      mAngleDiff -= Math.PI;
    } else if (mAngleDiff < -Math.PI / 2) {
      mAngleDiff += Math.PI;
    }
  }

  private void finish() {
    if (mInProgress) {
      mInProgress = false;
      if (mListener != null) {
        mListener.onRotationEnd(this);
      }
    }
  }

  public boolean onTouchEvent(MotionEvent event) {
    switch (event.getActionMasked()) {

      case MotionEvent.ACTION_DOWN:
        mInProgress = false;
        mPointerIndices[0] = event.getPointerId(event.getActionIndex());
        mPointerIndices[1] = MotionEvent.INVALID_POINTER_ID;
        break;

      case MotionEvent.ACTION_POINTER_DOWN:
        if (!mInProgress) {
          mPointerIndices[1] = event.getPointerId(event.getActionIndex());
          mInProgress = true;
          mPrevTime = event.getEventTime();
          mPrevAngle = Double.NaN;
          updateCurrent(event);
          if (mListener != null) {
            mListener.onRotationBegin(this);
          }
        }
        break;

      case MotionEvent.ACTION_MOVE:
        if (mInProgress) {
          updateCurrent(event);
          if (mListener != null) {
            mListener.onRotation(this);
          }
        }
        break;

      case MotionEvent.ACTION_POINTER_UP:
        if (mInProgress) {
          int pointerIndex = event.getPointerId(event.getActionIndex());
          if (pointerIndex == mPointerIndices[0] || pointerIndex == mPointerIndices[1]) {
            // One of the key pointer has been lifted up, we have to end the gesture
            finish();
          }
        }
        break;

      case MotionEvent.ACTION_UP:
        finish();
        break;
    }
    return true;
  }

  /**
   * Returns rotation in radians since the previous rotation event.
   *
   * @return current rotation step in radians.
   */
  public double getRotation() {
    return mAngleDiff;
  }

  /**
   * Return the time difference in milliseconds between the previous
   * accepted rotation event and the current rotation event.
   *
   * @return Time difference since the last rotation event in milliseconds.
   */
  public long getTimeDelta() {
    return mCurrTime - mPrevTime;
  }
}

File: android/lib/src/main/java/com/swmansion/gesturehandler/RotationGestureDetector.kt
package com.swmansion.gesturehandler

import android.view.MotionEvent
import kotlin.math.atan2

class RotationGestureDetector(private val mListener: OnRotationGestureListener?) {
  interface OnRotationGestureListener {
    fun onRotation(detector: RotationGestureDetector): Boolean
    fun onRotationBegin(detector: RotationGestureDetector): Boolean
    fun onRotationEnd(detector: RotationGestureDetector)
  }

  private var mCurrTime: Long = 0
  private var mPrevTime: Long = 0
  private var mPrevAngle = 0.0

  /**
   * Returns rotation in radians since the previous rotation event.
   *
   * @return current rotation step in radians.
   */
  var rotation = 0.0
    private set

  /**
   * Returns X coordinate of the rotation anchor point relative to the view that the provided motion
   * event coordinates (usually relative to the view event was sent to).
   *
   * @return X coordinate of the rotation anchor point
   */
  var anchorX = 0f
    private set

  /**
   * Returns Y coordinate of the rotation anchor point relative to the view that the provided motion
   * event coordinates (usually relative to the view event was sent to).
   *
   * @return Y coordinate of the rotation anchor point
   */
  var anchorY = 0f
    private set
  private var mInProgress = false
  private val mPointerIds = IntArray(2)
  private fun updateCurrent(event: MotionEvent) {
    mPrevTime = mCurrTime
    mCurrTime = event.eventTime
    val firstPointerIndex = event.findPointerIndex(mPointerIds[0])
    val secondPointerIndex = event.findPointerIndex(mPointerIds[1])
    val firstPtX = event.getX(firstPointerIndex)
    val firstPtY = event.getY(firstPointerIndex)
    val secondPtX = event.getX(secondPointerIndex)
    val secondPtY = event.getY(secondPointerIndex)
    val vectorX = secondPtX - firstPtX
    val vectorY = secondPtY - firstPtY
    anchorX = (firstPtX + secondPtX) * 0.5f
    anchorY = (firstPtY + secondPtY) * 0.5f

    // Angle diff should be positive when rotating in clockwise direction
    val angle = -atan2(vectorY.toDouble(), vectorX.toDouble())
    rotation = if (java.lang.Double.isNaN(mPrevAngle)) {
      0.0
    } else mPrevAngle - angle

    mPrevAngle = angle
    if (rotation > Math.PI) {
      rotation -= Math.PI
    } else if (rotation < -Math.PI) {
      rotation += Math.PI
    }
    if (rotation > Math.PI / 2.0) {
      rotation -= Math.PI
    } else if (rotation < -Math.PI / 2.0) {
      rotation += Math.PI
    }
  }

  private fun finish() {
    if (mInProgress) {
      mInProgress = false
      mListener?.onRotationEnd(this)
    }
  }

  fun onTouchEvent(event: MotionEvent): Boolean {
    when (event.actionMasked) {
      MotionEvent.ACTION_DOWN -> {
        mInProgress = false
        mPointerIds[0] = event.getPointerId(event.actionIndex)
        mPointerIds[1] = MotionEvent.INVALID_POINTER_ID
      }
      MotionEvent.ACTION_POINTER_DOWN -> if (!mInProgress) {
        mPointerIds[1] = event.getPointerId(event.actionIndex)
        mInProgress = true
        mPrevTime = event.eventTime
        mPrevAngle = Double.NaN
        updateCurrent(event)
        mListener?.onRotationBegin(this)
      }
      MotionEvent.ACTION_MOVE -> if (mInProgress) {
        updateCurrent(event)
        mListener?.onRotation(this)
      }
      MotionEvent.ACTION_POINTER_UP -> if (mInProgress) {
        val pointerId = event.getPointerId(event.actionIndex)
        if (pointerId == mPointerIds[0] || pointerId == mPointerIds[1]) {
          // One of the key pointer has been lifted up, we have to end the gesture
          finish()
        }
      }
      MotionEvent.ACTION_UP -> finish()
    }
    return true
  }

  /**
   * Return the time difference in milliseconds between the previous
   * accepted rotation event and the current rotation event.
   *
   * @return Time difference since the last rotation event in milliseconds.
   */
  val timeDelta: Long
    get() = mCurrTime - mPrevTime
}
